### PR TITLE
As a user, I can logout

### LINKED ISSUE #6: As a user, I can logout
## Why

To add the logout functionality so that user can logout after login

## Acceptance Criteria
- [x] Implement logout feature
- [x] Include unit test

### PR DESCRIPTION
**Pull request by junan**
https://github.com/junan/google-scraper/projects/1#card-50537726

## What happened 👀

- User can logout
- Refactored auth authorization to make it more structured  

### Logout

- Added boostrap navbar
- Changed routes mapping(`Get` => `New`, `Post` => `Create`)
- Added `Delete` action to handle logout
- Loaded boostarp and jquery from CDN

### Auth authorization

Created a struct to check access permission
```Go
type AuthPolicy struct {
	canAccess bool
}
```

Then setting the default access permission(User need to logged in to access any page) in the base controller

```Go
func (c *baseController) setDefaultAuthPolicy() {
	// By default user need to log in to access any routes.
	// Override this default policy in the `NestPrepare()` functions when necessary(ex: login, registration page)
	c.authPolicy = AuthPolicy{canAccess: c.isAuthenticatedUser()}
}
```

Finally override this default access permission when necessary(in the `NestPrepare` of child controller)

```Go
       //  In the registration controller where we want only guest user can access
	c.authPolicy.canAccess = c.isGuestUser()
```

## Proof Of Work 📹

RqsLAP1AgS

### CODE REVIEW

File: helpers/controller.go [hunk 1]
```diff
@@ -1,17 +1,31 @@
...
+	c.Data["CurrentUser"] = user
+	c.Data["navbarExpandModifierCssClass"] = navbarExpandModifierCssClass(user)
+	c.Data["AlertMap"] = map[string]string{
 		"success": "success",
-		"error": "danger",
+		"error":   "danger",
+	}
+}
+
+func navbarExpandModifierCssClass(user *models.User) string {
```
Comment (olivierobert) [inline helpers/controller.go]: Having complex logic for a class name does not seem the right way to implement the layout 🤔 It requires code in the controller, view and CSS styles 😓 Can you think fo an alternative even if it requires to change the UI?

Comment (junan) [reply helpers/controller.go]: @olivierobert I updated it by using two navbar partial(for guest user and autheticated user) in here https://github.com/junan/google-scraper/pull/27/commits/4132290e3b66006051c700ed9ff4aad6bf44e943

File: routers/router.go [hunk 2]
```diff
@@ -3,11 +3,14 @@ package routers
...
 func init() {
-	web.Router("/", &controllers.Dashboard{})
-	web.Router("/register", &controllers.Registration{})
-	web.Router("/login", &controllers.Session{})
+	web.Router("/", &controllers.Dashboard{}, "get:New")
+	web.Router("/register", &controllers.Registration{}, "get:New")
+	web.Router("/register", &controllers.Registration{}, "post:Create")
+	web.Router("/login", &controllers.Session{}, "get:New")
+	web.Router("/login", &controllers.Session{}, "post:Create")
+	web.Router("/logout", &controllers.Session{}, "get:Delete")
```
Comment (carryall) [inline routers/router.go]:
```suggestion
	web.Router("/register", &controllers.Registration{}, "get:New;post:Create")
	web.Router("/login", &controllers.Session{}, "get:New;post:Create")
	web.Router("/logout", &controllers.Session{}, "get:Delete")
```

File: helpers/controller.go (hunk 1 above)
Comment (olivierobert) [reply helpers/controller.go]: Simpler and easier to maintain 👏

File: views/layouts/authentication.html [hunk 3]
```diff
@@ -1,22 +1,19 @@
 <!DOCTYPE html>
```
Comment (olivierobert) [inline views/layouts/authentication.html]: Minor thing but this lie break is not needed 🙈

**Review by olivierobert — APPROVED** [review]

File: views/shared/authenticated_user_navbar.html [hunk 4]
```diff
@@ -0,0 +1,32 @@
...
+        <a class="nav-link" href="/">Dashboard</a>
+      </li>
+    </ul>
+
+    <ul class="navbar-nav ml-auto">
+      <li class="nav-item">
+        <span class="navbar-text">Logged in as {{ .CurrentUser.Name }}</span>
+      </li>
+
+      <li class="nav-item active">
```
Comment (carryall) [inline views/shared/authenticated_user_navbar.html]: @junan Not sure why does this have to include the `active` class 🤔

Comment (junan) [reply views/shared/authenticated_user_navbar.html]: `active` class makes the link text bright, which looks good to me. Thats why I added the class

Comment (carryall) [reply views/shared/authenticated_user_navbar.html]: @junan I mean normally we use `active` class on an active nav item so using it here just to style the link looks a bit weird to me 🤔  How about using another class here and set the color to that class?

**Review by carryall — APPROVED** [review]

**Pull request merged**